File: service/app/static/js/ha-events.js
/*
 * On-screen Home Assistant event channel (Pantry Raider).
 *
 * When enabled (Settings > Home Assistant), the page polls /events/poll for
 * events pushed from Home Assistant and shows them on the display:
 *   - notification -> a toast in the top-right, coloured by level
 *   - camera       -> a full-screen pop-up of a camera feed for a few seconds
 *
 * On load we read the current last id first, so a page that opens long after an
 * event was sent does not replay a backlog; we only show events that arrive
 * after the page connects. Pure vanilla JS with self-injected styles, so it has
 * no dependency on the rest of the page.
 */
(function () {
  var cfg = document.getElementById('ha-events-config');
  if (!cfg) return;
  // Per-device override (FoodAssistant-vcuz): a device can show or hide on-screen
  // events independently of the server default. localStorage 'haEventsShow' is
  // '1' (show), '0' (hide), or absent (follow the server default).
  var ov = null;
  try { ov = localStorage.getItem('haEventsShow'); } catch (e) { }
  // Whether Home Assistant on-screen events (notification toasts, camera
  // pop-ups, and the page changes HA pushes) show on this device. Deck-action
  // confirmations (FoodAssistant-rdlo) and device-health warnings
  // (FoodAssistant-h28s) ignore this: they are local feedback (a Stream Deck
  // press worked) or a local device alert (the Pi under-voltage), not Home
  // Assistant traffic, so they always show. That is why the script no longer
  // bails out when HA events are turned off.
  var enabled = ov === '1' ? true : (ov === '0' ? false : cfg.dataset.default === '1');

  // Should this event show now? Home Assistant events obey the toggle above; a
  // deck confirmation and a device-health warning always show, whether HA
  // events are on or off.
  function shouldShow(ev) {
    return !!ev && (enabled || ev.type === 'confirm' || ev.type === 'warning'
      || (ev.type === 'navigate' && ev.always));
  }

  // Events drive physical-control feedback: a NeoKey press jumps the kiosk to
  // the scan screen, and a button that takes even a fraction of a second feels
  // broken. So events LONG-POLL on their own channel: the request is held
  // server-side and returns the instant a press lands, with no fixed-tick lag,
  // instead of riding the slower consolidated status the heavier surfaces share.
  // A held request is cheap (an mtime-gated read of the in-memory ring), so this
  // stays light on a small Pi. Re-holds almost immediately on a good answer,
  // backs off when the server blips, and pauses while the tab is hidden.
  var WAIT_S = 25;             // how long the server holds an otherwise-idle poll
  var IDLE_MS = 50;            // tiny gap before re-holding, so a request is ~always open
  var BACKOFF_MAX_MS = 15000;
  var curDelay = IDLE_MS;
  var DEFAULT_TOAST_MS = 8000;
  var lastId = 0;
  var camTimer = null;
  var camSnapTimer = null;

  // --- styles + containers -------------------------------------------------
  var style = document.createElement('style');
  style.textContent =
    '.hae-toasts{position:fixed;top:64px;right:14px;z-index:2050;display:flex;flex-direction:column;gap:10px;max-width:360px}' +
    '.hae-toast{background:#1f242b;border:1px solid #333b45;border-left:5px solid #14b8c4;border-radius:10px;' +
    'padding:10px 36px 10px 12px;color:#e6e9ed;box-shadow:0 .4rem 1rem rgba(0,0,0,.4);position:relative;' +
    'opacity:0;transform:translateX(16px);transition:opacity .25s,transform .25s}' +
    '.hae-toast.show{opacity:1;transform:none}' +
    '.hae-toast .hae-title{font-weight:700;margin-bottom:2px}' +
    '.hae-toast .hae-x{position:absolute;top:6px;right:8px;cursor:pointer;color:#9aa4af;border:0;background:none;font-size:18px;line-height:1}' +
    '.hae-toast.info{border-left-color:#14b8c4}.hae-toast.success{border-left-color:#2ea043}' +
    '.hae-toast.warning{border-left-color:#d29922}.hae-toast.error{border-left-color:#da3633}' +
    '.hae-toast.hae-clickable:hover{background:#262c34}' +
    '.hae-cam{position:fixed;inset:0;z-index:2060;background:rgba(0,0,0,.92);display:flex;flex-direction:column;' +
    'align-items:center;justify-content:center;gap:12px}' +
    '.hae-cam img{max-width:96vw;max-height:82vh;object-fit:contain;background:#000;border-radius:8px}' +
    '.hae-cam .hae-cap{color:#e6e9ed;font-size:18px}' +
    '.hae-cam .hae-close{position:absolute;top:14px;right:18px;color:#fff;background:rgba(0,0,0,.5);border:0;' +
    'border-radius:8px;font-size:22px;padding:4px 12px;cursor:pointer}';
  document.head.appendChild(style);

  var toasts = document.createElement('div');
  toasts.className = 'hae-toasts';
  document.body.appendChild(toasts);

  // --- notification toast ---------------------------------------------------
  // A warning toast that names the settings pane where its fix lives
  // (FoodAssistant-44f6, e.g. a Pi temperature/under-voltage alert -> the
  // Network pane's device-health banner) is clickable: it deep-links there
  // instead of dropping the user on the generic /setup landing page. Already
  // on /setup, it just switches pane in place via openSettingsPane; anywhere
  // else it navigates to /setup#pane-... where setup/menu.js's init()
  // activates that hash on load.
  function goToPane(pane) {
    if (!pane) return;
    var onSetup = /(^|\/)setup(\.html)?$/.test(window.location.pathname);
    if (onSetup && typeof window.openSettingsPane === 'function') {
      window.openSettingsPane(pane);
    } else {
      window.location.assign('/setup#' + pane);
    }
  }
  function showToast(ev) {
    var el = document.createElement('div');
    var pane = ev.type === 'warning' ? String(ev.pane || '') : '';
    el.className = 'hae-toast ' + (ev.level || 'info') + (pane ? ' hae-clickable' : '');
    var title = ev.title ? '<div class="hae-title"></div>' : '';
    el.innerHTML = title + '<div class="hae-msg"></div><button class="hae-x" aria-label="Dismiss">&times;</button>';
    if (ev.title) el.querySelector('.hae-title').textContent = ev.title;
    el.querySelector('.hae-msg').textContent = ev.message || '';
    el.querySelector('.hae-x').onclick = function (e) { e.stopPropagation(); dismiss(el); };
    if (pane) {
      el.style.cursor = 'pointer';
      el.title = 'Open settings';
      el.onclick = function () { goToPane(pane); };
    }
    toasts.appendChild(el);
    requestAnimationFrame(function () { el.classList.add('show'); });
    var ms = (ev.timeout > 0 ? ev.timeout * 1000 : DEFAULT_TOAST_MS);
    setTimeout(function () { dismiss(el); }, ms);
  }
  function dismiss(el) {
    el.classList.remove('show');
    setTimeout(function () { if (el.parentNode) el.parentNode.removeChild(el); }, 260);
  }

  // --- camera pop-up --------------------------------------------------------
  function closeCamera() {
    if (camTimer) { clearTimeout(camTimer); camTimer = null; }
    if (camSnapTimer) { clearInterval(camSnapTimer); camSnapTimer = null; }
    var ex = document.querySelector('.hae-cam');
    if (ex && ex.parentNode) ex.parentNode.removeChild(ex);
  }
  function showCamera(ev) {
    if (!ev.src) return;
    closeCamera();
    var box = document.createElement('div');
    box.className = 'hae-cam';
    box.innerHTML = '<button class="hae-close" aria-label="Close">&times;</button>' +
      '<img alt="Camera"><div class="hae-cap"></div>';
    box.querySelector('.hae-cap').textContent = ev.name || 'Camera';
    box.querySelector('.hae-close').onclick = closeCamera;
    var img = box.querySelector('img');
    function refresh() {
      var sep = ev.src.indexOf('?') >= 0 ? '&' : '?';
      img.src = ev.src + sep + '_=' + Date.now();
    }
    document.body.appendChild(box);
    refresh();
    camSnapTimer = setInterval(refresh, 1500);
    var secs = ev.seconds > 0 ? ev.seconds : 20;
    camTimer = setTimeout(closeCamera, secs * 1000);
  }

  // --- navigate (HA drives the on-screen page) -----------------------------
  function doNavigate(ev) {
    var p = String(ev.path || '').replace(/^\/+/, '');
    if (!p) return;
    // Same-origin relative paths only: never follow a scheme or a
    // protocol-relative URL the server should already have refused.
    if (/^[a-z][a-z0-9+.-]*:/i.test(p) || p.indexOf('//') === 0) return;
    // A physical control drove this (a NeoKey press). Wake a blanked kiosk
    // right away rather than waiting for the page load to report activity, and
    // signal the current page so it can react even when we do not reload (a
    // mode changed while already on the scan screen: the mode tiles refresh
    // and bounce).
    try { if (window.__prKioskActivity) window.__prKioskActivity(); } catch (e) { }
    try { window.dispatchEvent(new Event('pr-nav-signal')); } catch (e) { }
    // Already on this page: don't reload (avoids a refresh loop).
    var here = window.location.pathname.replace(/^\/+/, '');
    if (here === p || here === p.split('?')[0]) return;
    try { window.location.assign(p); } catch (e) { /* ignore */ }
  }

  function handle(ev) {
    if (!shouldShow(ev)) return;
    // A confirmation and a device-health warning render through the same toast
    // path as a notification; they are just gated differently (always shown)
    // and coloured by their level: a confirm is success (green), a device
    // warning is warning (amber) or error (red), so a Pi alert reads clearly
    // apart from an "it worked".
    if (ev.type === 'notification' || ev.type === 'confirm' || ev.type === 'warning') showToast(ev);
    else if (ev.type === 'camera') showCamera(ev);
    else if (ev.type === 'navigate') doNavigate(ev);
  }

  // --- poll loop ------------------------------------------------------------
  // Fold one /events/poll answer (or the events slice of the consolidated
  // status) into the screen: advance the cursor and show each new event.
  function applyEvents(d) {
    if (!d) return;
    if (typeof d.last_id === 'number') lastId = Math.max(lastId, d.last_id);
    (d.events || []).forEach(handle);
  }

  function poll() {
    // A hidden tab has no screen to pop toasts or camera views on: skip the
    // long-poll and re-check slowly. Events are delivered by id (since=lastId),
    // so anything that arrives while hidden still shows on the next visible
    // poll rather than being lost.
    if (document.hidden) {
      setTimeout(poll, 2000);
      return;
    }
    // Held server-side until an event lands (or WAIT_S elapses), so a press is
    // delivered with ~no lag. Re-hold almost at once on any good answer; a blip
    // backs off so a down main server (a satellite whose forward failed) is not
    // hammered.
    fetch('events/poll?wait=' + WAIT_S + '&since=' + lastId, { cache: 'no-store' })
      .then(function (r) { return r.ok ? r.json() : null; })
      .then(function (d) {
        if (d) { curDelay = IDLE_MS; applyEvents(d); }
        else { curDelay = Math.min(Math.max(curDelay, IDLE_MS) * 2, BACKOFF_MAX_MS); }
      })
      .catch(function () { curDelay = Math.min(Math.max(curDelay, IDLE_MS) * 2, BACKOFF_MAX_MS); })
      .finally(function () { setTimeout(poll, curDelay); });
  }

  // Events get their own dedicated long-poll (see WAIT_S / IDLE_MS above): a
  // physical control's on-screen response has to feel immediate, and the
  // consolidated status poll the other surfaces share runs far slower (a 2s
  // floor, often 4s), which made a NeoKey press take seconds to move the
  // screen. Read the current last id first so a page opened long after an event
  // does not replay a backlog, then start the long-poll loop.
  fetch('events/poll?since=999999999', { cache: 'no-store' })
    .then(function (r) { return r.ok ? r.json() : null; })
    .then(function (d) { if (d && typeof d.last_id === 'number') lastId = d.last_id; })
    .catch(function () {})
    .finally(function () { setTimeout(poll, IDLE_MS); });
})();

File: service/app/services/ha_events.py
"""On-screen Home Assistant event channel.

Home Assistant pushes events to Pantry Raider (a rest_command in an automation),
and the kiosk / web UI polls for them and shows them on the display: notification
toasts and camera pop-ups (for example, pop up the doorbell camera when a person
is detected). Events live in a small ring capped by count and age, so a kiosk
that was off does not get flooded with a backlog when it polls.

Each event has a monotonically increasing ``id`` so a client can poll for "what
is new since the last id I saw" without missing or replaying events.

Sharing (FoodAssistant-0fho): the ring persists to a small state file under
data_dir, the same pattern as scanner_mode.py. A server running multiple
uvicorn workers must share the ring, or an event HA posts through one worker
never reaches the kiosk polling another worker. Reads check the file's mtime
and only re-parse when it changed, so a poll costs one stat call; polling
never writes the file, only adding an event does. If data_dir is not writable
(tests, a read-only mount) the module quietly degrades to the old
process-local in-memory behavior.
"""
from __future__ import annotations

import json
import os
import threading
import time
from pathlib import Path

# Keep only the most recent events, and drop anything older than the TTL, so a
# kiosk that was off does not get flooded with a backlog when it polls.
_MAX_EVENTS = 50
_TTL_SECONDS = 120

_lock = threading.Lock()
_events: list[dict] = []
_next_id = 1
# mtime of the state file our in-memory view corresponds to (None = never seen).
_mtime: int | None = None


def _state_file() -> Path:
    # Resolved per call (not at import) so tests that repoint data_dir work.
    from ..config import settings
    return Path(settings.data_dir) / "ha_events.json"


def _load_locked() -> None:
    """Refresh the in-process ring from the state file if it changed on disk.
    Caller holds the lock."""
    global _events, _next_id, _mtime
    try:
        sf = _state_file()
        mtime = sf.stat().st_mtime_ns
    except OSError:
        return  # no file yet (fresh install, or unwritable data_dir)
    if mtime == _mtime:
        return
    try:
        data = json.loads(sf.read_text())
        events = [e for e in data.get("events", []) if isinstance(e, dict) and "id" in e]
        nxt = int(data.get("next", 1))
    except (OSError, ValueError, TypeError):
        return  # a torn or corrupt file never breaks a poll; keep what we have
    _mtime = mtime
    _events = events
    _next_id = max(nxt, _next_id)


def _save_locked() -> None:
    """Write the ring to the state file (atomic replace, best effort). Caller
    holds the lock."""
    global _mtime
    sf = _state_file()
    try:
        tmp = sf.with_name(sf.name + ".tmp")
        tmp.write_text(json.dumps({"next": _next_id, "events": _events}))
        os.replace(tmp, sf)
        _mtime = sf.stat().st_mtime_ns
    except OSError:
        pass  # data_dir not writable: fall back to process-local behavior


def _prune_locked(now: float) -> None:
    global _events
    cutoff = now - _TTL_SECONDS
    _events = [e for e in _events if e["ts"] >= cutoff][-_MAX_EVENTS:]


def _add(event: dict) -> int:
    global _next_id
    now = time.time()
    with _lock:
        _load_locked()
        event["id"] = _next_id
        event["ts"] = now
        _next_id += 1
        _events.append(event)
        _prune_locked(now)
        _save_locked()
        return event["id"]


def add_notification(message: str, title: str = "", level: str = "info",
                     timeout: int = 0) -> int:
    """Queue a notification toast. ``level`` is info/success/warning/error."""
    lvl = level if level in ("info", "success", "warning", "error") else "info"
    return _add({
        "type": "notification",
        "message": str(message),
        "title": str(title),
        "level": lvl,
        "timeout": max(0, int(timeout or 0)),
    })


def add_confirmation(message: str, title: str = "", timeout: int = 5) -> int:
    """Queue a deck-action confirmation toast (FoodAssistant-rdlo).

    Distinct from an HA notification on purpose: the kiosk shows these even
    when on-screen Home Assistant events are turned off, because they are local
    feedback that a Stream Deck press worked, not Home Assistant traffic. Rides
    the same ring and since-id contract as everything else, so it degrades and
    shares across workers exactly like the rest."""
    return _add({
        "type": "confirm",
        "message": str(message),
        "title": str(title),
        "level": "success",
        "timeout": max(0, int(timeout or 0)),
    })


def add_warning(message: str, title: str = "", key: str = "",
                level: str = "warning", timeout: int = 0, pane: str = "") -> int:
    """Queue a device-health warning toast (FoodAssistant-h28s).

    Used for a live Raspberry Pi power/thermal condition (under-voltage,
    over-temperature, throttling). Distinct from a Home Assistant notification
    on purpose: like a deck confirmation it shows even when on-screen HA events
    are turned off, because it is a local device-health alert, not Home
    Assistant traffic. ``level`` is warning or error (anything else clamps to
    warning) so the toast reads as a real alert, not an "it worked". ``key`` is
    the underlying warning key (undervoltage, throttled, ...) so a client can
    tell one condition from another. ``pane`` is an optional settings pane id
    (e.g. "pane-network") the toast should deep-link to on click, instead of
    the generic /setup landing page (FoodAssistant-44f6); any warning can carry
    one, not just device health. Rides the same ring and since-id contract as
    everything else, so it degrades and shares across workers like the rest."""
    lvl = level if level in ("warning", "error") else "warning"
    return _add({
        "type": "warning",
        "message": str(message),
        "title": str(title),
        "key": str(key),
        "level": lvl,
        "timeout": max(0, int(timeout or 0)),
        "pane": str(pane or ""),
    })


def add_camera(name: str = "", src: str = "", seconds: int = 0) -> int:
    """Queue a camera pop-up. ``src`` is the proxy snapshot path the kiosk shows."""
    return _add({
        "type": "camera",
        "name": str(name),
        "src": str(src),
        "seconds": max(0, int(seconds or 0)),
    })


def add_navigate(path: str) -> int:
    """Queue a kiosk page-change event. ``path`` is an app-relative path (e.g.
    "ui/cook"), so a Home Assistant automation can drive which page the display
    shows (FoodAssistant-i4rs). The kiosk navigates same-origin only."""
    return _add({"type": "navigate", "path": str(path or "").strip()})


def poll(after_id: int = 0) -> dict:
    """Events newer than ``after_id``, plus the current last id.

    A fresh client should first read ``last_id`` (with after_id huge, or via the
    returned value) so it only sees events that arrive after it connects, rather
    than replaying the recent ring on load. Polling never writes the state file
    (the TTL/size prune is applied in memory and re-applied on the next add).
    """
    now = time.time()
    with _lock:
        _load_locked()
        _prune_locked(now)
        last = _next_id - 1
        try:
            after = int(after_id)
        except (TypeError, ValueError):
            after = 0
        fresh = [dict(e) for e in _events if e["id"] > after]
    return {"events": fresh, "last_id": last}


def last_id() -> int:
    with _lock:
        _load_locked()
        return _next_id - 1


def reset() -> None:
    """Clear all events and drop the state file (used by tests)."""
    global _events, _next_id, _mtime
    with _lock:
        _events = []
        _next_id = 1
        _mtime = None
        try:
            _state_file().unlink(missing_ok=True)
        except OSError:
            pass
